Fix integer overflow when set a large maximum value for atom table

sunboshan opened the pull request:
When setting maximum atom table size using `+t` option, there will be a integer overflow for a large size.

```
$ erl +t2147482625
ll_alloc: Cannot allocate 18446744073692774400 bytes of memory (of type "atom_tab").
```
The overflow is caused by the arithmetic operations on `int` type in the following line in [index.c](https://github.com/erlang/otp/blob/OTP-20.1.5/erts/emulator/beam/index.c#L61)
    
    Uint base_size = ((limit+INDEX_PAGE_SIZE-1)/INDEX_PAGE_SIZE)*sizeof(IndexSlot*);

We can use above example `2147482625` to illustrate how the overflow occurs.

```
expression                      decimal value   binary value (32 bits)              note
INDEX_PAGE_SIZE                 1024            0b00000000000000000000010000000000  defined in index.h
limit                           2147482625      0b01111111111111111111110000000001  user's input
limit+INDEX_PAGE_SIZE           -2147483647     0b10000000000000000000000000000001  overflow occurs here
last_result-1                   -2147483648     0b10000000000000000000000000000000
last_result/INDEX_PAGE_SIZE     -2097152        0b11111111111000000000000000000000  Two's complement division is using arithmetic right shift
last_result*sizeof(IndexSlot*)  -16777216       0b11111111000000000000000000000000  On 64-bit Erlang, size of pointer is 8 bytes
``` 

Before it assign `-16777216` to `base_size`, it needs to resize from `int` type to `long` type (`Uint` is `unsigned long` type defined in sys.h). Then it will change the type from `long` to `unsigned long`.

```
expression                    decimal value          binary value (64 bits)
last_result                   -1677216                                               0b11111111000000000000000000000000
(long) last_result            -1677216               0b1111111111111111111111111111111111111111000000000000000000000000
Uint base_size = last_result  18446744073692774400   0b1111111111111111111111111111111111111111000000000000000000000000
```
And that gives the error message above.

Note:
In [atom.h](https://github.com/erlang/otp/blob/OTP-20.1.5/erts/emulator/beam/atom.h#L39), the `MAX_ATOM_TABLE_SIZE` is `(UWORD_CONSTANT(1) << 32)`, namely `4294967296` for 64 bits Erlang. However, in [erl_init.c](https://github.com/erlang/otp/blob/OTP-20.1.5/erts/emulator/beam/erl_init.c#L2024), the user input is stored in a `long` type, however the [variable](https://github.com/erlang/otp/blob/OTP-20.1.5/erts/emulator/beam/erl_init.c#L208) holding it is a `int` type, so there would also be overflow happened here.

```
/* atom.h */
#define MAX_ATOM_TABLE_SIZE ((MAX_ATOM_INDEX + 1 < (UWORD_CONSTANT(1) << 32)) ? MAX_ATOM_INDEX + 1 : (UWORD_CONSTANT(1) << 32))

/* erl_init.c */
int erts_atom_table_size = ATOM_LIMIT;
...
erts_atom_table_size = strtol(arg, NULL, 10);  /* overflow occurs here, if user's input is MAX_ATOM_TABLE_SIZE 4294967296, then it will overflow to 0 */
```

Therefore even the intention for the maximum atom size is `1UL << 32 = 4294967296`, but the actual maximum is `(1 << 31) - 1 = 2147483647`(the maximum value a 4 bytes signed int can hold).

After the fix, it will behave like this:
```
$ erl +t2147483647
Erlang/OTP 21 [DEVELOPMENT] [erts-9.1.4] [source] [64-bit] [smp:4:4] [ds:4:4:10] [async-threads:10] [hipe]

Eshell V9.1.4  (abort with ^G)
1> erlang:system_info(atom_limit).
2147483647

$ erl +t2147483648
bad atom table size 2147483648
```

bjorng commented on erts/emulator/beam/index.c:
```diff
@@ -58,7 +58,7 @@ IndexTable*
 		int size, int limit, HashFunctions fun)
 {
-    Uint base_size = ((limit+INDEX_PAGE_SIZE-1)/INDEX_PAGE_SIZE)*sizeof(IndexSlot*);
+    Uint base_size = ((limit-1)/INDEX_PAGE_SIZE+1)*sizeof(IndexSlot*);
```
This does eliminate the integer overflow problem, but it will also for some values of `limit` unnecessarily round up one extra page. I suggest the following fix instead:
 
` Uint base_size = (((Uint)limit+INDEX_PAGE_SIZE-1)/INDEX_PAGE_SIZE)*sizeof(IndexSlot*);`

sunboshan commented:
Modified the fix as your suggestion, also added detailed description in commit message.

Regarding to the maximum atom size, on 64 bits Erlang it will be inconsistent between claimed max(4294967296) and actual max(2147483647). To fix it, either change the claimed max, or change the `erts_atom_table_size` to be `long`. However the latter will also require multiple places to be change, including the variable `limit` in the above `index.c`. Then it will require more change on various .h files.

BTW currently the helper message display the range of +t option is [8192-0].

```
$ erl +t0
bad atom table size 0
...
-t size        set the maximum number of atoms the emulator can handle
               valid range is [8192-0]
...
```

Due to [printf format](https://github.com/erlang/otp/blob/OTP-20.1.5/erts/emulator/beam/erl_init.c#L659-L660) is `[%d-%d]`, whereas [MAX_ATOM_TABLE_SIZE](https://github.com/erlang/otp/blob/OTP-20.1.5/erts/emulator/beam/atom.h#L39) is `unsigned long (1UL) << 32 = 4294967296 = 0b100000000000000000000000000000000 = 0(when format in %d)`. Fixing it would be changing printf format to `%ld`, but then this is not the real maximum value. So maybe leaving them all as is would be good for now? Since assuming no one will ever need more than 2147483647 atoms.

bjorng commented:
Good commit message! I only have to things I want changed: Remove the "." at the end of the first line of the first line, and squash the commits to one commit.

Regarding the helper message: It is OK to ignore for now. If you would want to fix it, you could do it in a separate commit in this PR, and lower the atom limit at same time as you fix the format.

sunboshan commented:
Changed commit messages, and also lower the MAX_ATOM_TABLE_SIZE to (1 << 31) - 1, since this is the actual max size, it will also correctly display the range of atom to be [8192-2147483647] on a 64-bits Erlang.

bjorng commented:
Thanks! Added to our daily builds.

bjorng commented:
No problems were found in our daily builds. We can merge this PR as soon as you have signed the Contributor License Agreement (click on Show all checks above).
